### PR TITLE
PP-426: Server periodic hooks infrastructure

### PR DESCRIPTION
**Pull request by arungrover**
#### Issue-ID

* **[PP-426](https://pbspro.atlassian.net/browse/PP-426)**

#### Problem description
* Add support for server periodic hooks

#### Cause / Analysis
* A new hook event needs to be added to run a hook in server periodically.

#### Solution description
* This part of the server periodic hook implementation only deals with accepting and launching a server periodic hook out of server process.
Design proposal is **[here](https://pbspro.atlassian.net/wiki/pages/viewpage.action?pageId=31621121)** 

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/include/work_task.h [hunk 1]
```diff
@@ -67,6 +67,11 @@ enum work_type {
 	                        */
 };
 
+enum wtask_delete_option {
+    DELETE_ONE,
+    DELETE_ALL
+};
```
Comment (arungrover) [inline src/include/work_task.h]: Well the enum is used only in a function which is meant to delete a work task... I can not think of a scenario where I would need a DELETE_NONE. But if you think it is for sanity purpose only then I can add it too.

File: test/tests/functional/pbs_server_periodic_hook.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        a = msg.rsplit(' ', 1)
+        return int(a[1])
+
+    def check_next_occurances(self, count, freq,
+                              hook_run_time, check_for_hook_end):
+        """
+        Helper function to check the occurances of hook by matching their
+        logs in server logs.
+        It needs 4 arguments:
+        - count			to know how many times to repeat
```
Comment (bhroam) [inline test/tests/functional/pbs_server_periodic_hook.py]: Is the indentation off?

File: test/tests/functional/pbs_server_periodic_hook.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                                checking these logs
+        - freq			is the frequency set in pbs server to run this hook
+        - hook_rum_time		is the amount of time hook takes to run.
```
Comment (bhroam) [inline test/tests/functional/pbs_server_periodic_hook.py]: is the indentation off?

**Comment by mike0042** [comment]
What is the status of this PR? It's getting a bit stale.

**Comment by arungrover** [comment]
@mike0042 I'm testing this change on a windows machine. Even though I do get a log message on windows from the new binary, It does not log anything from the actual hook.
When I change the logging in the hook to use pbs.logjobmsg instead of using pbs.logmsg, it starts to log.
All this is making me think that there is a logging problem on windows but I'm not sure.

So, testing is still in progress on this.

**Comment by mike0042** [comment]
Thank you for the update.

**Comment by bhroam** [comment]
I think this PR is ready to be merged.  Vinod approved the changes and mine were taken over by a new bug.  Please let me know and i'll merge it in.

**Comment by arungrover** [comment]
You are right @bhroam. I'll log the new ticket in JIRA about the logging as soon as it is up after maintenance.

Thanks!

**Comment by bhroam** [comment]
Thanks Arun, I merged in the PR.

**Pull request merged**